File: package/main/src/Time/convertTime.ts
import type { TimeUnit } from "$/time/timeUnit";
import type { TimeUnitShort } from "$/time/timeUnitShort";
import { division, multiplication } from "@/Math";
import { normalizeTimeUnit } from "@/Time/normalizeTimeUnit";

/**
 * Defines conversion rates between time units
 */
const conversionRates: Record<TimeUnit, number> = {
  milliseconds: 1,
  seconds: 1000,
  minutes: 60_000,
  hours: 3_600_000,
};

/**
 * Converts time between different units
 * @param value Value to convert (string or number)
 * @param fromUnit Source time unit
 * @param toUnit Target time unit
 * @returns Converted value (number)
 * @throws {Error} If the input value is invalid
 */
export const convertTime = (
  value: string | number,
  fromUnit: TimeUnit | TimeUnitShort,
  toUnit: TimeUnit | TimeUnitShort,
): number => {
  const numericValue =
    typeof value === "string" ? Number.parseFloat(value) : value;
  const normalizedFromUnit = normalizeTimeUnit(fromUnit, "long");
  const normalizedToUnit = normalizeTimeUnit(toUnit, "long");
  const milliseconds = multiplication(
    numericValue,
    conversionRates[normalizedFromUnit],
  );
  return division(milliseconds, conversionRates[normalizedToUnit]);
};

File: package/umt_python/src/time/convert_time.py
from typing import Literal, Union
from .normalize_time_unit import normalize_time_unit

TimeUnit = Literal["milliseconds", "seconds", "minutes", "hours"]
TimeUnitShort = Literal["ms", "s", "m", "h"]

CONVERSION_RATES = {
    "milliseconds": 1,
    "seconds": 1000,
    "minutes": 60000,
    "hours": 3600000,
}


def convert_time(
    value: Union[str, int, float],
    from_unit: Union[TimeUnit, TimeUnitShort],
    to_unit: Union[TimeUnit, TimeUnitShort],
) -> float:
    """
    Converts time between different units.

    Args:
        value: Value to convert (string or number)
        from_unit: Source time unit
        to_unit: Target time unit

    Returns:
        Converted value (number)

    Example:
        >>> convert_time(1, "hours", "minutes")
        60.0
        >>> convert_time("1000", "ms", "s")
        1.0
    """
    numeric_value = float(value) if isinstance(value, str) else value
    normalized_from_unit = normalize_time_unit(from_unit, "long")
    normalized_to_unit = normalize_time_unit(to_unit, "long")

    milliseconds = numeric_value * CONVERSION_RATES[normalized_from_unit]
    return milliseconds / CONVERSION_RATES[normalized_to_unit]
